File: sections/35_math_mixed_contest/problems/b_coprime_pairs/solution.py
import sys
from collections import defaultdict


def spf_table(n: int) -> list[int]:
    spf = list(range(n + 1))
    if n >= 1:
        spf[1] = 1
    for p in range(2, n + 1):
        if spf[p] == p and p * p <= n:
            for x in range(p * p, n + 1, p):
                if spf[x] == x:
                    spf[x] = p
    return spf


def prime_factors(x: int, spf: list[int]) -> list[int]:
    factors = []
    while x > 1:
        p = spf[x]
        factors.append(p)
        while x % p == 0:
            x //= p
    return factors


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    n = data[0]
    arr = data[1 : 1 + n]
    spf = spf_table(max(arr))
    seen = defaultdict(int)
    answer = 0
    for idx, x in enumerate(arr):
        factors = prime_factors(x, spf)
        share = 0
        m = len(factors)
        for mask in range(1, 1 << m):
            prod = 1
            bits = 0
            for i, p in enumerate(factors):
                if (mask >> i) & 1:
                    prod *= p
                    bits += 1
            share += seen[prod] if bits % 2 else -seen[prod]
        answer += idx - share
        for mask in range(1, 1 << m):
            prod = 1
            for i, p in enumerate(factors):
                if (mask >> i) & 1:
                    prod *= p
            seen[prod] += 1
    print(answer)


if __name__ == "__main__":
    main()

File: sections/35_math_mixed_contest/problems/b_coprime_pairs/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    if (!(cin >> n)) return 0;
    vector<int> a(n);
    int mx = 1;
    for (int &x : a) {
        cin >> x;
        mx = max(mx, x);
    }
    vector<int> spf(mx + 1);
    iota(spf.begin(), spf.end(), 0);
    if (mx >= 1) spf[1] = 1;
    for (long long p = 2; p <= mx; p++) {
        if (spf[p] == p && p * p <= mx) {
            for (long long x = p * p; x <= mx; x += p) {
                if (spf[x] == x) spf[x] = p;
            }
        }
    }
    vector<int> seen(mx + 1, 0);
    long long answer = 0;
    for (int idx = 0; idx < n; idx++) {
        int x = a[idx];
        vector<int> factors;
        while (x > 1) {
            int p = spf[x];
            factors.push_back(p);
            while (x % p == 0) x /= p;
        }
        long long share = 0;
        int m = factors.size();
        for (int mask = 1; mask < (1 << m); mask++) {
            int prod = 1, bits = 0;
            for (int i = 0; i < m; i++) if (mask & (1 << i)) {
                prod *= factors[i];
                bits++;
            }
            if (bits % 2) share += seen[prod];
            else share -= seen[prod];
        }
        answer += idx - share;
        for (int mask = 1; mask < (1 << m); mask++) {
            int prod = 1;
            for (int i = 0; i < m; i++) if (mask & (1 << i)) prod *= factors[i];
            seen[prod]++;
        }
    }
    cout << answer << '\n';
    return 0;
}
